### PR TITLE
Allow passing in container_root for extra_vars file

### PR DESCRIPTION
**Pull request by fosterseth**
Allow passing in argument for the container_root (default is /runner)

In OCP/K8S, projects run in the task pod's ee container. The private_data_dir/ is not extracted to /runner/, instead project update runs directly from the mounted in private_data_dir, e.g. `/tmp/awx_1_blahblah`

When setting up extra vars for certain project types (like Insights), we need to pass in `-e @/tmp/awx_1_blahblah/env/tmpextravarsfile`

Without this change, it always uses `/runner/env/tmpextravarsfile`, which is not correct and leads to a project failure

### CODE REVIEW

File: src/awx_plugins/interfaces/_temporary_private_container_api.py [hunk 1]
```diff
@@ -39,7 +40,10 @@ def get_incontainer_path(
     if not os.path.isabs(private_data_dir):
         raise RuntimeError('The private_data_dir path must be absolute')
 
-    container_root = pathlib.Path(CONTAINER_ROOT)
+    if not container_root:
+        container_root = CONTAINER_ROOT
+
+    container_root = pathlib.Path(container_root)
```
Comment (webknjaz) [inline src/awx_plugins/interfaces/_temporary_private_container_api.py]: (nit) this could be shorter
```suggestion
    container_root = pathlib.Path(container_root or CONTAINER_ROOT)
```

File: src/awx_plugins/interfaces/_temporary_private_inject_api.py [hunk 2]
```diff
@@ -159,6 +159,7 @@ def inject_credential(
     safe_env: EnvVarsType,
     args: ArgsType,
     private_data_dir: str,
+    **kwargs: dict[str, str],
```
Comment (webknjaz) [inline src/awx_plugins/interfaces/_temporary_private_inject_api.py]: Let's avoid kwargs when possible:
```suggestion
    *,
    container_root: str = '',
```

Comment (fosterseth) [reply src/awx_plugins/interfaces/_temporary_private_inject_api.py]: I think you forgot to drop the `*,`

Comment (webknjaz) [reply src/awx_plugins/interfaces/_temporary_private_inject_api.py]: I didn't. I kept what seemed to be your intent.

File: src/awx_plugins/interfaces/_temporary_private_inject_api.py [hunk 3]
```diff
@@ -287,7 +288,10 @@ def inject_credential(
         )
         if extra_vars:
             path = _build_extra_vars_file(extra_vars, private_data_dir)
-            container_path = get_incontainer_path(path, private_data_dir)
+            container_root = kwargs.get('container_root', '')
+            container_path = get_incontainer_path(
+                path, private_data_dir, container_root=container_root,
```
Comment (webknjaz) [inline src/awx_plugins/interfaces/_temporary_private_inject_api.py:295]: The MyPy error here (reported @ https://results.pre-commit.ci/run/github/858143066/1741025635.F2wg0bCpRjSZ-U2dUypT0w) is caused by using `kwargs.get()` since its type declaration says that it can basically return `typing.Any` which is not `str`. And so it doesn't like letting you pass arbitrarily-typed things where only strings should fit.

Specifically, it's defined in https://github.com/python/typeshed/blob/6f98c59/stdlib/builtins.pyi#L1144-L1150. If kwargs were to be a typed dict, it'd narrow the type properly.

File: src/awx_plugins/interfaces/_temporary_private_container_api.py [hunk 4]
```diff
@@ -18,6 +18,7 @@
 def get_incontainer_path(
         path: os.PathLike[str] | str,
         private_data_dir: os.PathLike[str] | str,
+        container_root: str = '',
```
Comment (webknjaz) [inline src/awx_plugins/interfaces/_temporary_private_container_api.py]: I missed this earlier, but this could actually be anything path-like since we're converting it through `pathlib.Path()` anyway. The argument types should be wide, while the return value should be typed narrowly.
```suggestion
        container_root: os.PathLike[str] | str = '',
```

File: src/awx_plugins/interfaces/_temporary_private_container_api.py [hunk 5]
```diff
@@ -39,7 +40,7 @@ def get_incontainer_path(
     if not os.path.isabs(private_data_dir):
         raise RuntimeError('The private_data_dir path must be absolute')
 
-    container_root = pathlib.Path(CONTAINER_ROOT)
+    container_root = pathlib.Path(container_root or CONTAINER_ROOT)
```
Comment (webknjaz) [inline src/awx_plugins/interfaces/_temporary_private_container_api.py]: Oh, this causes a type confusion. This is because the incoming argument type is `str` and not `str | Path` and we're assigning an incompatible type here. Just reassigning doesn't override the implied type definition.

I recommend using a new variable for the new bit of data.
```suggestion
    container_root_path = pathlib.Path(container_root or CONTAINER_ROOT)
```

File: src/awx_plugins/interfaces/_temporary_private_container_api.py [hunk 6]
```diff
@@ -39,7 +40,7 @@ def get_incontainer_path(
     if not os.path.isabs(private_data_dir):
         raise RuntimeError('The private_data_dir path must be absolute')
 
-    container_root = pathlib.Path(CONTAINER_ROOT)
+    container_root_path = pathlib.Path(container_root or CONTAINER_ROOT)
```
Comment (webknjaz) [inline src/awx_plugins/interfaces/_temporary_private_container_api.py]: (don't forget to adjust variable name in the return statement)

File: src/awx_plugins/interfaces/_temporary_private_container_api.py [hunk 7]
```diff
@@ -18,6 +18,7 @@
 def get_incontainer_path(
         path: os.PathLike[str] | str,
         private_data_dir: os.PathLike[str] | str,
+        container_root: os.PathLike[str] | str = '',
```
Comment (webknjaz) [inline src/awx_plugins/interfaces/_temporary_private_container_api.py]: There's a discussion to be had about this API. Should the new argument allowed to be positional? As the number of arguments grow, it's usually best to disallow positional args because it hurts readability and ease of perception on the calling side.

File: src/awx_plugins/interfaces/_temporary_private_inject_api.py [hunk 8]
```diff
@@ -287,7 +289,10 @@ def inject_credential(
         )
         if extra_vars:
             path = _build_extra_vars_file(extra_vars, private_data_dir)
-            container_path = get_incontainer_path(path, private_data_dir)
+            container_root = kwargs.get('container_root', '')
```
Comment (webknjaz) [inline src/awx_plugins/interfaces/_temporary_private_inject_api.py]: Dropping this should make MyPy happy.
```suggestion
```

File: src/awx_plugins/interfaces/_temporary_private_container_api.py (hunk 7 above)
Comment (fosterseth) [reply src/awx_plugins/interfaces/_temporary_private_container_api.py]: is now keyword only.

File: src/awx_plugins/interfaces/_temporary_private_container_api.py (hunk 6 above)
Comment (fosterseth) [reply src/awx_plugins/interfaces/_temporary_private_container_api.py]: fixed it up

File: src/awx_plugins/interfaces/_temporary_private_container_api.py [hunk 9]
```diff
@@ -18,6 +18,8 @@
 def get_incontainer_path(
         path: os.PathLike[str] | str,
         private_data_dir: os.PathLike[str] | str,
+        *,
+        container_root: os.PathLike[str] | str = '',
```
Comment (webknjaz) [inline src/awx_plugins/interfaces/_temporary_private_container_api.py]: @fosterseth would it make sense to just have
```suggestion
        container_root: os.PathLike[str] | str = pathlib.Path(CONTAINER_ROOT),
```
here instead of that fallback later on?

File: src/awx_plugins/interfaces/_temporary_private_inject_api.py (hunk 3 above)
Comment (webknjaz) [reply src/awx_plugins/interfaces/_temporary_private_inject_api.py:295]: Marking this as resolved.

File: src/awx_plugins/interfaces/_temporary_private_container_api.py (hunk 9 above)
Comment (webknjaz) [reply src/awx_plugins/interfaces/_temporary_private_container_api.py]: Or maybe have a `None` sentinel instead?
```suggestion
        container_root: os.PathLike[str] | str | None = None,
```

I realized that having an empty string as a magic value for "maybe nothing was passed into the argument" isn't the best look. Semantically, a dedicated sentinel would be better, or a sane default.

File: src/awx_plugins/interfaces/_temporary_private_inject_api.py [hunk 10]
```diff
@@ -159,6 +159,8 @@ def inject_credential(
     safe_env: EnvVarsType,
     args: ArgsType,
     private_data_dir: str,
+    *,
+    container_root: str = '',
```
Comment (webknjaz) [inline src/awx_plugins/interfaces/_temporary_private_inject_api.py]: Same concern as in the other module: an empty string sentinel should be avoided. Plus the type here is string, but it's better to accept a wider range of types in function args.

File: src/awx_plugins/interfaces/_temporary_private_container_api.py [hunk 11]
```diff
@@ -39,15 +43,18 @@ def get_incontainer_path(
     if not os.path.isabs(private_data_dir):
         raise RuntimeError('The private_data_dir path must be absolute')
 
-    container_root = pathlib.Path(CONTAINER_ROOT)
+    container_root_path = pathlib.Path(container_root or CONTAINER_ROOT)
```
Comment (webknjaz) [inline src/awx_plugins/interfaces/_temporary_private_container_api.py]: Did you decide against defaulting to `pathlib.Path(CONTAINER_ROOT)`?
If `None` is used as a sentinel, the correct check would be
```suggestion
    container_root_path = pathlib.Path(CONTAINER_ROOT if container_root is None else container_root)
```

**Comment by webknjaz** [comment]
@fosterseth please add test coverage for both modified functions that goes through the code paths with `None`, strings and pathlib objects, maybe some corner cases...

File: tests/_temporary_private_container_api_test.py [hunk 12]
```diff
@@ -87,3 +87,58 @@ def test_paths_outside_private_path_conversion(
...
+    (pathlib.Path, str),
+    ids=('host-path-pathlib', 'host-path-str'),
+)
+@pytest.mark.parametrize(
+    'convert_runner_path_to_type',
+    (pathlib.Path, str),
+    ids=('host-path-pathlib', 'host-path-str'),
+)
+# pylint: disable-next=too-many-arguments,too-many-positional-arguments
+def test_provide_custom_container_root(  # noqa: WPS211
```
Comment (webknjaz) [inline tests/_temporary_private_container_api_test.py:127]: @fosterseth could you look into adding similar coverage for `None` vs. path for `inject_credential`?

**Review by webknjaz — APPROVED** [review]
LGTM

**Review by webknjaz — APPROVED** [review]

**Comment by AlanCoding** [comment]
> In OCP/K8S, projects run in the task pod's ee container. The private_data_dir/ is not extracted to /runner/, instead project update runs directly from the mounted in private_data_dir, e.g. /tmp/awx_1_blahblah

This is the best explanation for the high-level thing going on here, but it still doesn't add up, so let me talk that out.

We mount the project directory in normal installs and run over receptor too. It runs in podman in that case. The volume mount is, in no way, unique to OCP/K8S.

In addition to that, project credentials are attached via the `scm_credential`, not the generalized `credentials`. If I take things at face value, it seems that someone may have reversed this, specifically for the insights credential type, and then encountered errors because of that. The AWX patch modifies the _main_ job credential injection path. Getting clear on how we're even in that code path for a project update is not fully established for me.

Even assuming these details are all correct (doubtful), I don't this is a good solution to the problem. It seems the simpler solution is to either change where we are volume mounting in OCP or in traditional project updates so we do not have a special case.
